Library/Homebrew/dev-cmd/pr-pull.rb

```diff
@@ -0,0 +1,96 @@
+# frozen_string_literal: true
+
+require "cli/parser"
+require "utils/github"
+require "tmpdir"
+
+module Homebrew
+  module_function
+
+  def pr_pull_args
+    Homebrew::CLI::Parser.new do
+      usage_banner <<~EOS
+        `pr-pull` <pull_request>
+
+        Download and publish bottles, and apply the bottle commit from a
+        pull request with artifacts generated from GitHub Actions.
+        Requires write access to the repository.
+      EOS
+      switch "--no-publish",
+             description: "Download the bottles, apply the bottle commit, and "\
+                          "upload the bottles to Bintray, but don't publish them."
+      switch "--no-upload",
+             description: "Download the bottles and apply the bottle commit, "\
+                          "but don't upload to Bintray."
+      switch "--dry-run", "-n",
+             description: "Print what would be done rather than doing it."
+      switch "--clean",
+             description: "Do not amend the commits from pull requests."
+      switch "--workflow=",
+             description: "Specify the workflow name (default: tests.yml)."
+      switch :verbose
+      switch :debug
+      min_named 1
+    end
+  end
+
+  def signoff!(pr, path: ".", dry_run: false)
+    message = `git -C #{path} log -1 --pretty=%B`
+    close_message = "Closes ##{pr}."
+    message += "\n#{close_message}" unless message.include? close_message
+    if dry_run
+      puts "git commit --amend --signoff -m $message"
+    else
+      safe_system "git", "-C", path, "commit", "--amend", "--signoff", "--allow-empty", "-q", "-m", message
+    end
+  end
+
+  def cherry_pick_pr!(pr, path: ".", dry_run: false)
+    if dry_run
+      puts <<~EOS
+        git fetch --force origin +refs/pull/#{pr}/head
+        git merge-base HEAD FETCH_HEAD
+        git cherry-pick $merge_base..FETCH_HEAD
+      EOS
+    else
+      safe_system "git", "-C", path, "fetch", "--quiet", "--force", "origin", "+refs/pull/#{pr}/head"
+      merge_base = `git -C #{path} merge-base HEAD FETCH_HEAD`.strip
+      safe_system "git", "-C", path, "cherry-pick", "#{merge_base}..FETCH_HEAD"
+    end
+  end
+
+  def pr_pull
+    pr_pull_args.parse
+
+    if !ENV["HOMEBREW_BINTRAY_USER"] || !ENV["HOMEBREW_BINTRAY_KEY"]
+      odie "Missing HOMEBREW_BINTRAY_USER or HOMEBREW_BINTRAY_KEY variables!" unless args.dry_run? || args.no_upload?
+    end
+
+    workflow_name = args.workflow || "tests.yml"
+    test_bot_cmd = %W[#{HOMEBREW_PREFIX}/bin/brew test-bot --ci-upload]
```

Comment:
Not super happy about a Homebrew/brew command that calls `brew test-bot`. `test-bot` is a bit of a mess (even moreso than `brew pull`). Would be nice to instead perhaps move the upload code from `test-bot` into Homebrew/brew and have it call back. In a perfect world: it'd be a separate command that `brew test-bot` can call (but this command calls the methods rather than the command) to be able to have the workflow be a bit more modular if stuff goes wrong.

Reply:
I've added a new `bintray.rb` that implements the Bintray API. `brew test-bot` can later be updated to use these new functions.